internal/cache/safedemands.go

```
@@ -0,0 +1,148 @@
+	t := time.NewTicker(time.Minute)
+	defer t.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-t.C:
+			if sdc.checkDemandCRDExists(ctx) {
+				return nil
+			}
+		}
+	}
+}
+
+func (sdc *SafeDemandCache) checkDemandCRDExists(ctx context.Context) bool {
+	_, ready, err := crd.CheckCRDExists(demandapi.DemandCustomResourceDefinitionName(), sdc.apiExtensionsClient)
+	if err != nil {
+		svc1log.FromContext(ctx).Info("failed to determine if demand CRD exists", svc1log.Stacktrace(err))
+		return false
+	}
+	if ready {
+		svc1log.FromContext(ctx).Info("demand CRD has been initialized. Demand resources can now be created")
+		err = sdc.initializeCache(ctx)
+		if err != nil {
+			svc1log.FromContext(ctx).Error("failed initializing demand cache", svc1log.Stacktrace(err))
+			return false
+		}
+	}
+	return ready
+}
+
+func (sdc *SafeDemandCache) initializeCache(ctx context.Context) error {
+	sdc.cacheInitialization.Lock()
+	defer sdc.cacheInitialization.Unlock()
+	if sdc.demandCRDInitialized.Load() {
+		return nil
+	}
+	informer := sdc.informerFactory.Scaler().V1alpha1().Demands()
+	sdc.informerFactory.Start(ctx.Done())
+
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, 2*time.Second)
+	defer cancel()
+	if ok := clientcache.WaitForCacheSync(ctxWithTimeout.Done(), informer.Informer().HasSynced); !ok {
+		return werror.Error("timeout syncing informer", werror.SafeParam("timeoutSeconds", 2))
+	}
+	demandCache, err := NewDemandCache(informer, sdc.demandKubeClient)
+	if err != nil {
+		return err
+	}
+	demandCache.Run(ctx)
+	sdc.DemandCache = demandCache
+	sdc.demandCRDInitialized.Store(true)
+	return nil
+}
+
+func (sdc *SafeDemandCache) IsLoaded() bool {
+	return sdc.demandCRDInitialized.Load()
+}
+
+// Create enqueues a creation request and puts the object into the store
+func (sdc *SafeDemandCache) Create(rr *demandapi.Demand) error {
+	if !sdc.demandCRDInitialized.Load() {
+		return nil
```

should return an error